File: web/src/lib/idle_events.js
// Idle events: when a voice call goes quiet, draw one of the companion's
// configured events by weight and hand its prompt to the companion as a
// hidden note — a check-in after a silence, a topic to riff on, a message
// from the stream's chat to answer. The server stores the list
// (agents.idle_events, cleaned by server/idle_events.py) and sends the
// usable events with the call's start payload; the clock lives here,
// because only the browser knows when the companion's audio actually
// finished playing and when the user last spoke or typed.
//
// Timing:
//   - the quiet stretch counts from the end of PLAYBACK, not from when the
//     reply text arrives, so an event never fires over a line still being
//     spoken;
//   - with a min-max range, every quiet stretch re-rolls its length, so
//     events don't land like clockwork;
//   - a streak of events nobody answers pauses them until someone speaks,
//     types or chats. That is also what lets the idle-hangup watchdog end a
//     forgotten call: every companion turn counts as call activity.
//   - silent chat injections ('silent_chat') sit outside the draw: after
//     their own quiet time the newest chat is queued as background for the
//     companion's next reply, without asking for one.
import { rpc } from "./rpc";
import { _t } from "./i18n";
import { notification } from "./notification";

// Keep in sync with server/idle_events.py NOTE_PREFIX: the saved copy of a
// note is a system row the transcript hides by this prefix.
const NOTE_PREFIX = "[Idle event";
const TICK_MS = 500;
// Chat state poll while the companion has a chat event — the Minecraft
// pump's cadence. Each poll also keeps the server's chat readers running.
const CHAT_POLL_MS = 4000;
const PLATFORM_LABELS = { twitch: "Twitch", youtube: "YouTube" };

/** A chat block: one heading naming the platform ("Stream chat" only when a
 *  read mixes Twitch and YouTube), then a "[name]: message" line each - the
 *  group-call relay's speaker format. */
function chatLines(chat) {
    const platforms = [...new Set(chat.map((m) => m.platform))];
    return [
        platforms.length === 1 ? `${PLATFORM_LABELS[platforms[0]] || platforms[0]} chat:` : "Stream chat:",
        ...chat.map((m) => `[${m.user}]: ${m.text}`),
    ];
}

/** 1 → "1st", 2 → "2nd", 11 → "11th", 23 → "23rd". */
function ordinal(n) {
    const tens = n % 100;
    if (tens >= 11 && tens <= 13) return `${n}th`;
    return `${n}${{ 1: "st", 2: "nd", 3: "rd" }[n % 10] || "th"}`;
}

export class IdleEventScheduler {
    constructor(call) {
        this.call = call;   // VoiceCallService
        this.cfg = null;
        this._timer = null;
    }

    /** Arm for this call from the primary leg's start payload (null = the
     *  companion has idle events off). A compaction restart re-arms too. */
    configure(cfg) {
        this.stop();
        if (!cfg?.events?.length) return;
        this.cfg = cfg;
        this._drawn = cfg.events.filter((e) => e.type !== "silent_chat");
        this._silent = cfg.events.filter((e) => e.type === "silent_chat");
        this._usesChat = cfg.events.some((e) => e.type !== "prompt");
        this._injectedThisQuiet = false;
        this._quietSince = Date.now();
        this._quietTarget = null;
        this._unanswered = 0;
        this._capLogged = false;
        this._lastEvent = null;
        this._userSeq = 0;
        this._firing = false;
        this._chat = null;
        this._chatPolledAt = 0;
        this._warned = new Set();
        this._timer = setInterval(() => this._tick(), TICK_MS);
        console.log(`[idle_events] armed: ${cfg.events.length} event(s), `
            + `${cfg.min_seconds}-${cfg.max_seconds} s of quiet, `
            + `pause after ${cfg.max_unanswered || "∞"} unanswered`);
    }

    stop() {
        if (this._timer) clearInterval(this._timer);
        this._timer = null;
        this.cfg = null;
    }

    /** The user spoke or typed: the quiet stretch starts over and the
     *  unanswered streak resets. */
    noteUser() {
        if (!this.cfg) return;
        this._userSeq += 1;
        this._unanswered = 0;
        this._restartQuiet();
    }

    _restartQuiet() {
        this._quietSince = Date.now();
        this._quietTarget = null;
        this._injectedThisQuiet = false;
    }

    /** Anything that holds an event back, on any leg of the call: a reply
     *  being generated or still playing, a tool round owing its follow-up,
     *  the user mid-utterance, a compaction restart. */
    _busy() {
        return [...this.call.connections.values()].some((c) => !c.isTerminal && (
            c._responseInFlight || c._pendingToolReply || c._toolReplyStarting
            || c._assistantAudioActive() || c._userSpeaking || c.state.compacting
            || c.toolDispatcher?.hasPending?.()));
    }

    _tick() {
        if (!this.cfg || this.call.state.status !== "live") return;
        if (this._usesChat && Date.now() - this._chatPolledAt >= CHAT_POLL_MS) this._pollChat();
        if (this._firing) return;
        if (this._busy()) {
            this._restartQuiet();
            return;
        }
        const quietMs = Date.now() - this._quietSince;
        // Silent chat injections keep their own quiet time, outside the
        // weighted draw and the unanswered pause (they ask for no reply).
        if (this._silent.length && !this._injectedThisQuiet) this._injectChat(quietMs);
        if (!this._drawn.length || this._firing) return;
        const { min_seconds: lo, max_seconds: hi, max_unanswered: cap } = this.cfg;
        if (cap > 0 && this._unanswered >= cap) {
            if (!this._capLogged) {
                this._capLogged = true;
                console.log(`[idle_events] ${cap} in a row went unanswered — paused until someone speaks, types or chats`);
            }
            return;
        }
        this._capLogged = false;
        if (this._quietTarget == null) this._quietTarget = (lo + Math.random() * (hi - lo)) * 1000;
        if (quietMs >= this._quietTarget) this._fire(quietMs);
    }

    /** Weighted draw among the events that can run now — chat events only
     *  while there is unread chat. A plain prompt doesn't fire twice in a
     *  row while another event could; chat events bring fresh messages
     *  each time, so they may repeat. */
    _pick() {
        const chatReady = (this._chat?.unread || 0) > 0;
        let pool = this._drawn.filter((e) => e.type !== "stream_chat" || chatReady);
        if (this._lastEvent?.type === "prompt" && pool.length > 1) {
            pool = pool.filter((e) => e !== this._lastEvent);
        }
        const total = pool.reduce((sum, e) => sum + e.weight, 0);
        let r = Math.random() * total;
        for (const e of pool) {
            r -= e.weight;
            if (r < 0) return e;
        }
        return null;
    }

    async _fire(quietMs) {
        const event = this._pick();
        if (!event) {
            // Only chat events, and nothing unread: wait out another stretch.
            this._restartQuiet();
            return;
        }
        this._firing = true;
        const userSeq = this._userSeq;
        try {
            let chat = [];
            if (event.type === "stream_chat") {
                const res = await rpc("/api/live_chat/take", { count: event.chat_messages });
                chat = res?.messages || [];
                if (this._chat) this._chat.unread = 0;   // a read empties the unread pool
                if (!chat.length) return;   // aged out, or a moderator cleared it meanwhile
            }
            // The floor may have changed hands while chat was fetched.
            if (!this.cfg || userSeq !== this._userSeq || this._busy()
                || this.call.state.status !== "live") return;
            const quietS = Math.round(quietMs / 1000);
            // How many nudges in a row nobody has answered, so the companion
            // can tell a fresh silence from a long one.
            const streak = this._unanswered
                ? `, ${ordinal(this._unanswered + 1)} in a row with no reply` : "";
            const lines = [`${NOTE_PREFIX} — quiet for ${quietS} s${streak}] ${event.prompt}`];
            if (chat.length) lines.push(...chatLines(chat));
            const note = lines.join("\n");
            const primary = this.call.primary;
            if (!primary.injectContextItem(note, { promptResponse: true })) return;
            // Also saved, as a hidden system row: a resumed conversation
            // still knows what the companion was answering.
            primary.recordMessage({ role: "system", content: note });
            this._lastEvent = event;
            this._unanswered += 1;
            console.log(`[idle_events] "${event.name || event.prompt.slice(0, 40)}" after ${quietS} s of quiet`
                + (chat.length ? ` with ${chat.length} chat message(s)` : "")
                + ` (${this._unanswered} unanswered)`);
        } catch (e) {
            console.warn("[idle_events] event failed", e);
        } finally {
            this._firing = false;
            this._restartQuiet();
        }
    }

    /** A silent chat injection: the newest unread messages go in as
     *  background for the companion's next reply, once per quiet stretch,
     *  without asking for one. Queued rather than sent now - the voice API
     *  hides text sent ahead of the user's next spoken turn from the reply
     *  to it (see VoiceCallService.queueSilentContext). */
    async _injectChat(quietMs) {
        const event = this._silent.find((e) => quietMs >= e.after_seconds * 1000);
        if (!event || !((this._chat?.unread || 0) > 0)) return;
        this._injectedThisQuiet = true;
        this._firing = true;
        try {
            const res = await rpc("/api/live_chat/take", { count: event.chat_messages });
            const chat = res?.messages || [];
            if (this._chat) this._chat.unread = 0;   // a read empties the unread pool
            if (!chat.length || !this.cfg || this.call.state.status !== "live") return;
            const note = [`${NOTE_PREFIX} — chat in the background]${event.prompt ? ` ${event.prompt}` : ""}`,
                ...chatLines(chat)].join("\n");
            this.call.queueSilentContext(note);
            this.call.primary.recordMessage({ role: "system", content: note });
            console.log(`[idle_events] "${event.name || "silent chat"}" queued ${chat.length} chat message(s) for the next reply`);
        } catch (e) {
            console.warn("[idle_events] chat injection failed", e);
        } finally {
            this._firing = false;
        }
    }

    async _pollChat() {
        this._chatPolledAt = Date.now();
        let state;
        try {
            state = await rpc("/api/live_chat/state", {});
        } catch (e) {
            return;   // server restarting — the next poll catches up
        }
        if (!this.cfg) return;
        // New chat counts as someone answering: a stream with a lively chat
        // never runs into the unanswered pause.
        if (this._chat && state.received > this._chat.received) this._unanswered = 0;
        this._chat = state;
        const platforms = Object.entries(state.platforms || {});
        if (!platforms.length) {
            this._warnOnce("setup", _t("This companion has chat idle events, but no live chat is set up — add a Twitch channel or a YouTube stream in Settings → Live chat."));
        }
        for (const [platform, p] of platforms) {
            if (p.status === "error" && p.error) {
                this._warnOnce(platform, _t("%s chat: %s", PLATFORM_LABELS[platform] || platform, p.error));
            }
        }
    }

    _warnOnce(key, message) {
        if (this._warned.has(key)) return;
        this._warned.add(key);
        notification.add(message, { type: "warning" });
    }
}

File: server/idle_events.py
# Copyright 2026 Codemarchant
"""Idle events: prompts a companion gets when a voice call goes quiet.

The browser runs the clock (web/src/lib/idle_events.js) - it is the side
that knows when the companion's audio finished playing and when the user
last spoke or typed. The server stores each companion's list (the
agents.idle_events JSON column), cleans it on save, and hands the usable
events to a call at session start. Chat-reading events draw on the live
chat feed in live_chat.py."""
import json

# Every event note starts with this (web/src/lib/idle_events.js NOTE_PREFIX).
# The note is saved as a system row so a resumed conversation still knows
# what the companion was answering; session_service._transcript_rows keeps
# it out of the visible transcript by this prefix.
NOTE_PREFIX = '[Idle event'

# What an event does: 'prompt' = its prompt, drawn by weight in a quiet
# stretch; 'stream_chat' = the same plus the newest unread stream-chat
# messages (live_chat.take); 'silent_chat' = not drawn - after its own
# after_seconds of quiet the newest unread chat goes in as background for
# the companion's next reply, without asking for one.
TYPES = ('prompt', 'stream_chat', 'silent_chat')

_MAX_EVENTS = 50
# A stream-chat event takes at most this many of the newest unread
# messages - live_chat's whole unread pool.
_MAX_CHAT_MESSAGES = 50
_MAX_NAME = 80
_MAX_PROMPT = 4000


def _int(value, lo, hi, fallback):
    try:
        return max(lo, min(hi, int(value)))
    except (TypeError, ValueError):
        return fallback


def clean_events(value):
    """The idle_events list in its stored shape: {name, prompt, weight,
    type, chat_messages, after_seconds, active} dicts. Accepts the list or its JSON text; malformed
    entries are dropped rather than refused (imports carry whatever another
    install wrote). Weights are relative - 100 is the editor's default, so a
    lone event reads as 100%."""
    if isinstance(value, str):
        try:
            value = json.loads(value or '[]')
        except ValueError:
            value = []
    if not isinstance(value, list):
        return []
    out = []
    for item in value[:_MAX_EVENTS]:
        if not isinstance(item, dict):
            continue
        out.append({
            'name': str(item.get('name') or '')[:_MAX_NAME],
            'prompt': str(item.get('prompt') or '')[:_MAX_PROMPT],
            'weight': _int(item.get('weight', 100), 0, 1000, 100),
            'type': item.get('type') if item.get('type') in TYPES else 'prompt',
            # 'stream_chat' only: how many of the newest unread messages ride
            # along. Kept for the other types too, so switching back doesn't
            # lose the number.
            'chat_messages': _int(item.get('chat_messages', 10), 1, _MAX_CHAT_MESSAGES, 10),
            # 'silent_chat' only: seconds of quiet before the chat goes in.
            'after_seconds': _int(item.get('after_seconds', 0), 0, 3600, 0),
            'active': bool(item.get('active', True)),
        })
    return out


def events_json(value):
    """JSON text for the column. Compact separators and this key order match
    the editor's JSON.stringify, so a list the user didn't touch never reads
    as an unsaved edit after a reload."""
    return json.dumps(clean_events(value), ensure_ascii=False, separators=(',', ':'))


def call_payload(agent):
    """What a voice call needs to run the idle clock, or None when the
    companion has idle events off or no event that could fire (inactive,
    or a drawn event with a blank prompt or zero weight). Clamped here as
    well as in the editor: imported packages skip the editor."""
    if not agent['idle_events_enabled']:
        return None
    events = [e for e in clean_events(agent['idle_events'])
              if e['active'] and (e['type'] == 'silent_chat'
                                  or (e['prompt'].strip() and e['weight'] > 0))]
    if not events:
        return None
    lo = _int(agent['idle_events_min_seconds'], 1, 3600, 10)
    return {
        'min_seconds': lo,
        'max_seconds': _int(agent['idle_events_max_seconds'], lo, 3600, lo),
        'max_unanswered': _int(agent['idle_events_max_unanswered'], 0, 1000, 6),
        'events': [{'name': e['name'], 'prompt': e['prompt'].strip(), 'weight': e['weight'],
                    'type': e['type'], 'chat_messages': e['chat_messages'],
                    'after_seconds': e['after_seconds']} for e in events],
    }
